subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/ResolvedVariant.java

```diff
@@ -38,4 +39,6 @@
     AttributeContainerInternal getAttributes();
 
     ResolvedArtifactSet getArtifacts();
+
+    CapabilitiesMetadata getCapabilities();
```

Comment:
I think I may not fully grok how this all interacts or where this ends up surfacing in the public APIs...

How is this CapabilitiesMetadata different from the one available through VariantResolveMetadata.Identifier?

Are we going to have CapabilitiesMetadata in cases where VariantResolveMetadata.Identifier is null now? Does that make sense?

Reply:
That case is mainly for file based dependencies. They usually have at least the `artifactType` attribute and no capability. So in that case the capabilities will be an empty container.

Note that `VariantResolveMetadata.Identifier` does not have capabilities information, this is available on the `VariantResolveMetadata` itself.

As for the surfacing in the public API, these changes only impact what is show through `ResolvedVariantResult.capabilities` which only shows the list of capabilities and not the container.

Reply:
And the information is needed on this type because this is how transforms carry the variant information around.